### DISCUSSION TITLE
Best way to make/test changes to this core when using platformio

### DISCUSSION THREAD

**Discussion by LinusHeu** [earlephilhower/arduino-pico#1529] [Q&A] [answered]
I'm very much a beginner at using PlatformIO.
What's the best way to:
1. try out any commit/branch of this core?
2. make changes to this core, try them out and upload them to github? (Fork repo, clone to my PC, modify it, use that in Platformio, upload to github.)

What I tried:
1. Adding `platform_packages = framework-arduinopico@https://github.com/earlephilhower/arduino-pico.git` to platformio.ini
This successfully pulls the latest changes from master. However, I can't really edit it locally and commit my changes. Git says "This repository has no remotes configured to fetch from." 
And changing it to a different branch, version, etc causes PlatformIO to re-download the entire core which takes more than 15 minutes.

2. Creating a fork of this repo, git cloning it to a local folder and adding `platform_packages = framework-arduinopico@file://C:\Users\linus\Projekte\arduino-pico` to platformio.ini. 
"PlatformIO: Configuring project: Tool Manager: Installing file://C:\Users\linus\Projekte\arduino-pico" needs more than 15 minutes to install it from a folder on my blazing-fast SSD. And PlatformIO doesn't detect it when I do any changes to the core in that source folder. Not even when I click Dependencies->Update.

Sorry, I know, this is more of a PlatformIO than a specific arduino-pico question. But I didn't really find anything on google. (Except these 2 options that I tried and don't work that well.)

**Comment by earlephilhower** [comment]
Ping @maxgerhardt if you have a second on this question....

**Comment by maxgerhardt** [comment, marked as answer]
1. Fork Arduino-Pico to your Github account (or gitlab or wherever)
2.  **Recursively** download your Arduino-Pico fork somewhere. (`git clone --recursive <url>`)
3. Point PlatformIO to use the **symlink://** (ZERO COPY) protocol, not `file://` protocol, using a `platformio.ini` instruction of e.g.
```ini
platform_packages =
   framework-arduinopico@symlink://C:\Users\linus\Projekte\arduino-pico
```
4. Add the forked Arduino Pico folder to your VSCode workspace (just drag+drop it into your VSCode File explorer --> Add Folder To Workspace)
5. Do the changes in Arduino-Pico
6. `git add ; git commit -m "Message" ; git push` in the arduino-pico folder as needed
7. From your fork: Click "Contribute" button or go in this repo -> Pull Requests -> new -> Compare Across Forks -> Your fork into this main repo.

Note: If you add new board variants and definitions and want to test them in PlatformIO, follow what's written in https://github.com/earlephilhower/arduino-pico/issues/1532#issuecomment-1592755427

**Reply by LinusHeu** [reply]
Thank you very much.
What happens when submodules are updated? Does it automatically download and checkout the correct versions of the submodules when I do a pull UI in VS Code? Or do I have to do something like --recursive again?

**Reply by maxgerhardt** [reply]
A regular `git pull` will not update the submodules, that needs a `git submodules update --recursive`.
